Add sha256sum for every python package, and supporting code to check it before installation

berdario opened the pull request:
This is actually a very important change.

I wouldn't feel confident in letting friends use pythonz without it.

The problem is that without this change we blindly trust whatever we have downloaded, and execute the code contained on that package on our system.

Even if the download is going over https, if the remote server gets compromised (and similar things happened already with other open source projects) or if the user has extraneous trusted root CA installed on their machine (due e.g. to a crappy manufacturer like Lenovo: think of Superfish) all bets are off.

Moreover, the stackless packages are even going through plain text HTTP.

Every package manager worth its salt does the same (dpkg/apt, Nix... and also pythonz's competition: pyenv)

I took some hashes from the pyenv project, updated others, and I manually calculated the hashes for a few versions that pyenv doesn't handle.

I tested a few samples, but I haven't tested every single version in pythonz yet

saghul commented:
THis one looks really useful, thanks!

berdario commented:
Thank you for having a look at all the PR/Issues so soon!

saghul commented on pythonz/downloader.py:
```diff
@@ -79,3 +81,13 @@ def fetch(cls, url, filename):
+
+def validate_sha256(filename, sha256sum):
+    if sha256sum is not None:
+        with open(filename, 'rb') as f, mmap(f.fileno(), 0, access=ACCESS_READ) as m:
```
I get an expection here:

~~~~
Traceback (most recent call last):
  File "pythonz/commands/install.py", line 97, in run_command
    p.install()
  File "pythonz/installer/pythoninstaller.py", line 198, in install
    self.download_and_extract()
  File "pythonz/installer/pythoninstaller.py", line 217, in download_and_extract
    self.download()
  File "pythonz/installer/pythoninstaller.py", line 107, in download
    if not validate_sha256(self.download_file, self.expected_sha256):
  File "pythonz/downloader.py", line 88, in validate_sha256
    with open(filename, 'rb') as f, mmap(f.fileno(), 0, access=ACCESS_READ) as m:
AttributeError: __exit__
~~~~

berdario replied:
Uhm, you're probably using it with an old Python version.
Can't remember when mmap has been made a context manager... The workaround for old pythons is to wrap it in `contextlib.closing`
-- 
Sent from mobile. Please excuse my brevity.

saghul replied:
I'm using 2.7.10.

saghul commented on pythonz/downloader.py:
```diff
@@ -79,3 +81,13 @@ def fetch(cls, url, filename):
+        with open(filename, 'rb') as f, mmap(f.fileno(), 0, access=ACCESS_READ) as m:
+            return sha256(m).hexdigest() == sha256sum
+    else:
+        logger.warning('sha256sum unavailable, skipping verification.\nMake '
```
Please use the same type of quoting here and in the line below.

berdario commented:
I added the fix for Python2 (And I also added a small smoke [test in pew](https://github.com/berdario/pew/blob/master/tests/test_install.py) to ensure that it works on every version)

saghul commented:
Sorry for the delay, I was on vacation :sunglasses: Rebased and landed in 13b7fca, thanks a lot!
